# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/remoteworkers/RemoteWorkerChild.cpp
# Commit: f2debe329bca
# Full Hash: f2debe329bcaaad1248a234c165ccf240d602344
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2025-04-15 21:21:06
# Regressor Bug: 1899503
# File Overlap Count: 4
# Description:
#   Bug 1899503 - P3 PRemoteWorkerDebugger binding implementation. r=asuth
#   
#   This patch implements the PRemoteWorkerDebugger binding.
#   
#   We are trying to keep the same logic debugger registration/unregistration to minimize the difference between the local debugger(the debugger on the content process main thread)
# ==============================================================================

diff -r 604c4c52574a -r f2debe329bca dom/workers/remoteworkers/RemoteWorkerChild.cpp
--- a/dom/workers/remoteworkers/RemoteWorkerChild.cpp	Tue Apr 15 09:47:39 2025 +0000
+++ b/dom/workers/remoteworkers/RemoteWorkerChild.cpp	Tue Apr 15 09:47:39 2025 +0000
@@ -139,7 +139,8 @@
     : mState(VariantType<remoteworker::Pending>(), "RemoteWorkerState"),
       mServiceKeepAlive(RemoteWorkerService::MaybeGetKeepAlive()),
       mIsServiceWorker(aData.serviceWorkerData().type() ==
-                       OptionalServiceWorkerData::TServiceWorkerData) {
+                       OptionalServiceWorkerData::TServiceWorkerData),
+      mPendingOps("PendingRemoteWorkerOps") {
   MOZ_ASSERT(RemoteWorkerService::Thread()->IsOnCurrentThread());
 }
 
@@ -813,9 +814,34 @@
   }
 }
 
+void RemoteWorkerChild::PendRemoteWorkerOp(RefPtr<RemoteWorkerOp> aOp) {
+  MOZ_ASSERT_DEBUG_OR_FUZZING(mIsThawing);
+
+  auto pendingOps = mPendingOps.Lock();
+
+  pendingOps->AppendElement(std::move(aOp));
+}
+
+void RemoteWorkerChild::RunAllPendingOpsOnMainThread() {
+  RefPtr<RemoteWorkerChild> self = this;
+
+  auto pendingOps = mPendingOps.Lock();
+
+  for (auto& op : *pendingOps) {
+    op->StartOnMainThread(self);
+  }
+
+  pendingOps->Clear();
+}
+
 void RemoteWorkerChild::MaybeStartOp(RefPtr<RemoteWorkerOp>&& aOp) {
   MOZ_ASSERT(aOp);
 
+  if (mIsThawing) {
+    PendRemoteWorkerOp(std::move(aOp));
+    return;
+  }
+
   auto lock = mState.Lock();
 
   if (!aOp->MaybeStart(this, lock.ref())) {